### PR TITLE
Refining the Bitmask

### LINKED ISSUE #9698: Refining Bitmask
### What would you like to share?

Need to edit some of the changes in the Bitmask.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by imSanko**
### Describe your change:

Refined Bitmark

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #9698".

### CODE REVIEW

**Comment by imSanko** [comment]
@cclauss Can you take a look I've just added a few stuff

File: dynamic_programming/bitmask.py [hunk 1]
```diff
@@ -11,75 +11,54 @@
...
 
 
-class AssignmentUsingBitmask:
-    def __init__(self, task_performed, total):
-        self.total_tasks = total  # total no of tasks (N)
-
-        # DP table will have a dimension of (2^M)*N
-        # initially all values are set to -1
+class TaskAssignment:
+    def __init__(self, task_performed, total_tasks):
```
Comment (cclauss) [inline dynamic_programming/bitmask.py:15]: No type hint, no doctests?

**Comment by imSanko** [comment]
@cclauss can you have a glance and let me know

File: dynamic_programming/bitmask.py (hunk 1 above)
Comment (cclauss) [reply dynamic_programming/bitmask.py:15]: This is not resolved.  Please read CONTRIBUTING.md.  Functions need type hints and doctests.

**Review by cclauss — CHANGES_REQUESTED** [review]

**Comment by cclauss** [comment]
To fix the `pre-commit errors` go to https://github.com/imSanko/Python and click the `Sync fork` and then `Update branch` buttons.

File: dynamic_programming/bitmask.py (hunk 1 above)
Comment (cclauss) [inline dynamic_programming/bitmask.py:15]:
```suggestion
    def __init__(self, task_performed: list[list[int]], total_tasks: int) -> None:
```

File: dynamic_programming/tempCodeRunnerFile.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        task_performed (List[List[int]]): List of lists where each inner list represents
+            the tasks that can be performed by a person.
+        total_tasks (int): The total number of tasks to be assigned.
+        """
+        self.total_tasks = total_tasks
+        self.dp = [[-1 for _ in range(total_tasks + 1)] for _ in range(1 << len(task_performed))]
+        self.task: Dict[int, List[int]] = defaultdict(list)
+        self.final_mask: int = (1 << len(task_performed)) - 1
+
+    def count_ways_until(self, mask: int, task_no: int) -> int:
```
[the remark this answers is not in the document]
Comment (imSanko) [reply dynamic_programming/tempCodeRunnerFile.py:33]: from collections import defaultdict
from typing import List, Dict

class TaskAssignment:
    def __init__(self, task_performed: List[List[int]], total_tasks: int) -> None:
        """
        Initialize the TaskAssignment class.

        Parameters:
        task_performed (List[List[int]]): List of lists where each inner list represents
            the tasks that can be performed by a person.
        total_tasks (int): The total number of tasks to be assigned.
        """
        self.total_tasks = total_tasks
        self.dp = [[-1 for _ in range(total_tasks + 1)] for _ in range(1 << len(task_performed))]
        self.task: Dict[int, List[int]] = defaultdict(list)
        self.final_mask: int = (1 << len(task_performed)) - 1

    def count_ways_until(self, mask: int, task_no: int) -> int:
        """
        Recursively count the number of ways to assign tasks to people.

        Parameters:
        mask (int): Bitmask representing the tasks assigned to people.
        task_no (int): The current task number being considered.

        Returns:
        int: The number of ways to assign tasks.

        Doctest:
        >>> assignment = TaskAssignment([[1, 2], [2, 3]], 3)
        >>> assignment.count_ways_until(0, 1)
        4
        """
        if mask == self.final_mask:
            return 1
        if task_no > self.total_tasks:
            return 0
        if self.dp[mask][task_no] != -1:
            return self.dp[mask][task_no]

        total_ways_util: int = self.count_ways_until(mask, task_no + 1)

        if task_no in self.task:
            for person in self.task[task_no]:
                if mask & (1 << person):
                    continue
                total_ways_util += self.count_ways_until(mask | (1 << person), task_no + 1)

        self.dp[mask][task_no] = total_ways_util
        return self.dp[mask][task_no]

    def count_total_ways(self, task_performed: List[List[int]]) -> int:
        """
        Count the total number of ways to distribute tasks to people.

        Parameters:
        task_performed (List[List[int]]): List of lists where each inner list represents
            the tasks that can be performed by a person.

        Returns:
        int: The total number of ways to distribute tasks.

        Doctest:
        >>> assignment = TaskAssignment([[1, 2], [2, 3]], 3)
        >>> assignment.count_total_ways([[1], [2], [3]])
        2
        """
        for person, tasks in enumerate(task_performed):
            for task in tasks:
                self.task[task].append(person)

        return self.count_ways_until(0, 1)

if __name__ == "__main__":
    total_tasks = 5
    task_performed = [[1, 3, 4], [1, 2, 5], [3, 4]]
    
    assignment = TaskAssignment(task_performed, total_tasks)
    total_ways = assignment.count_total_ways(task_performed)
    
    print("Total number of ways to distribute tasks:", total_ways)

**Review by imSanko — COMMENTED** [review]
from collections import defaultdict
from typing import List, Dict

class TaskAssignment:
    def __init__(self, task_performed: List[List[int]], total_tasks: int) -> None:
        """
        Initialize the TaskAssignment class.

        Parameters:
        task_performed (List[List[int]]): List of lists where each inner list represents
            the tasks that can be performed by a person.
        total_tasks (int): The total number of tasks to be assigned.
        """
        self.total_tasks = total_tasks
        self.dp = [[-1 for _ in range(total_tasks + 1)] for _ in range(1 << len(task_performed))]
        self.task: Dict[int, List[int]] = defaultdict(list)
        self.final_mask: int = (1 << len(task_performed)) - 1

    def count_ways_until(self, mask: int, task_no: int) -> int:
        """
        Recursively count the number of ways to assign tasks to people.

        Parameters:
        mask (int): Bitmask representing the tasks assigned to people.
        task_no (int): The current task number being considered.

        Returns:
        int: The number of ways to assign tasks.

        Doctest:
        >>> assignment = TaskAssignment([[1, 2], [2, 3]], 3)
        >>> assignment.count_ways_until(0, 1)
        4
        """
        if mask == self.final_mask:
            return 1
        if task_no > self.total_tasks:
            return 0
        if self.dp[mask][task_no] != -1:
            return self.dp[mask][task_no]

        total_ways_util: int = self.count_ways_until(mask, task_no + 1)

        if task_no in self.task:
            for person in self.task[task_no]:
                if mask & (1 << person):
                    continue
                total_ways_util += self.count_ways_until(mask | (1 << person), task_no + 1)

        self.dp[mask][task_no] = total_ways_util
        return self.dp[mask][task_no]

    def count_total_ways(self, task_performed: List[List[int]]) -> int:
        """
        Count the total number of ways to distribute tasks to people.

        Parameters:
        task_performed (List[List[int]]): List of lists where each inner list represents
            the tasks that can be performed by a person.

        Returns:
        int: The total number of ways to distribute tasks.

        Doctest:
        >>> assignment = TaskAssignment([[1, 2], [2, 3]], 3)
        >>> assignment.count_total_ways([[1], [2], [3]])
        2
        """
        for person, tasks in enumerate(task_performed):
            for task in tasks:
                self.task[task].append(person)

        return self.count_ways_until(0, 1)

if __name__ == "__main__":
    total_tasks = 5
    task_performed = [[1, 3, 4], [1, 2, 5], [3, 4]]
    
    assignment = TaskAssignment(task_performed, total_tasks)
    total_ways = assignment.count_total_ways(task_performed)
    
    print("Total number of ways to distribute tasks:", total_ways)

File: dynamic_programming/bitmask.py [hunk 3]
```diff
@@ -9,77 +9,85 @@
 Find the total no of ways in which the tasks can be distributed.
 """
 from collections import defaultdict
+from typing import List, Dict
```
Comment (cclauss) [inline dynamic_programming/bitmask.py:12]: These are not needed in Python 3.12 (or with `from __future__ import annotations`), just use `list` and `dict`.

File: dynamic_programming/bitmask.py [hunk 4]
```diff
@@ -9,77 +9,85 @@
...
 
 
-class AssignmentUsingBitmask:
-    def __init__(self, task_performed, total):
-        self.total_tasks = total  # total no of tasks (N)
+class TaskAssignment:
+    def __init__(self, task_performed: List[List[int]], total_tasks: int) -> None:
```
Comment (cclauss) [inline dynamic_programming/bitmask.py:16]:
```suggestion
    def __init__(self, task_performed: list[list[int]], total_tasks: int) -> None:
```

File: dynamic_programming/bitmask.py [hunk 5]
```diff
@@ -9,77 +9,85 @@
...
+        """
+        Initialize the TaskAssignment class.
 
-        # DP table will have a dimension of (2^M)*N
-        # initially all values are set to -1
+        Parameters:
+        task_performed (List[List[int]]): List of lists where each inner list represents
```
Comment (cclauss) [inline dynamic_programming/bitmask.py:21]: Please do not repeat the datatypes that are stated above and checked by mypy.  Readers get confused if one changes and the other does not.
```suggestion
        task_performed: List of lists where each inner list represents
```

**Pull request closed without merging**